File: backend/src/repositories/in-memory-track.repository.ts
import { LoopTrackMetadata } from '../models/track';
import { TrackRepository } from './track.repository';

export class InMemoryTrackRepository implements TrackRepository {
  private tracks = new Map<string, LoopTrackMetadata>();

  async listTracks() {
    return Array.from(this.tracks.values()).sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
  }

  async getTrackById(trackId: string) {
    return this.tracks.get(trackId) ?? null;
  }

  async createTrack(track: LoopTrackMetadata) {
    this.tracks.set(track.id, track);

    return track;
  }

  async reset() {
    this.tracks.clear();
  }
}

export const trackRepository = new InMemoryTrackRepository();

File: backend-dotnet/src/Loopr.Api/Repositories/InMemoryTrackRepository.cs
using Loopr.Api.Domain;

namespace Loopr.Api.Repositories;

public sealed class InMemoryTrackRepository : ITrackRepository
{
    private readonly Dictionary<string, LoopTrackMetadata> tracks = new(StringComparer.Ordinal);
    private readonly object syncRoot = new();

    public Task<IReadOnlyList<LoopTrackMetadata>> ListAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            return Task.FromResult<IReadOnlyList<LoopTrackMetadata>>(
                tracks.Values.OrderByDescending(track => track.UpdatedAt).ToArray()
            );
        }
    }

    public Task<LoopTrackMetadata?> GetByIdAsync(
        string trackId,
        CancellationToken cancellationToken
    )
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            tracks.TryGetValue(trackId, out var track);

            return Task.FromResult(track);
        }
    }

    public Task<LoopTrackMetadata> CreateAsync(
        LoopTrackMetadata track,
        CancellationToken cancellationToken
    )
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            tracks[track.Id] = track;
        }

        return Task.FromResult(track);
    }

    public Task ResetAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            tracks.Clear();
        }

        return Task.CompletedTask;
    }
}
